README.rst
Kinto Redis
############

|travis| |coveralls|

.. |travis| image:: https://github.com/Kinto/kinto-redis/actions/workflows/test.yml/badge.svg
    :target: https://github.com/Kinto/kinto-redis/actions

.. |coveralls| image:: https://coveralls.io/repos/github/Kinto/kinto-redis/badge.svg?branch=master
    :target: https://coveralls.io/github/Kinto/kinto-redis?branch=master

**Kinto Redis** is a redis driver for `Kinto <https://kinto.readthedocs.io>`_
storage, permissions and cache backends.

.. note::

   The backend currently doesn't support transaction and will not work
   with plugins that are using the ResourceChanged event to stop the
   user action. i.e To validate the request or to handle quota management.

Installing ``kinto-redis``
==========================

You can use PyPI either installing kinto redis dependencies::

    pip install kinto[redis]

Or installing kinto-redis directly::

    pip install kinto-redis


Using Kinto Redis backends
==========================

After installing the ``kinto-redis`` package using PyPI, you can
configure your server like that::

    #
    # Backends.
    #
    # https://kinto.readthedocs.io/en/latest/configuration/settings.html#storage
    #
    kinto.storage_backend = kinto_redis.storage
    kinto.storage_url = redis://localhost:6379/1
    kinto.cache_backend = kinto_redis.cache
    kinto.cache_url = redis://localhost:6379/2
    kinto.permission_backend = kinto_redis.permission
    kinto.permission_url = redis://localhost:6379/3


Running the tests
=================

To run the unit tests::

  $ make tests


CHANGELOG.rst
Changelog
=========


2.2.0 (unreleased)
------------------

- Nothing changed yet.


2.1.0 (2021-11-16)
------------------

**Internal Changes**

- Replacing ``ujson`` with ``kinto.core.utils.json``


2.0.1 (2019-05-24)
------------------

**Bug fixes**

- ``pool_size`` setting should remain optional


2.0.0 (2019-02-08)
------------------

**Breaking changes**

- Upgrade to Kinto >= 13 storage API.


1.3.0 (2018-04-26)
------------------

- Update storage tests with new Kinto 9.x features. (#13)


1.2.0 (2017-08-17)
------------------

- Cache set now requires a ttl value
- Cache delete() method now returns the deleted value
- Cache never accept to store bytes.


1.1.0 (2017-02-23)
------------------

- Upgrade to last storage, permissions and cache backends features. (#7)


1.0.1 (2016-08-18)
------------------

**Bug fixes**

- Fix compability with Kinto 4.0 about unique fields (Kinto/kinto#763)


1.0.0 (2016-08-11)
------------------

- Move the kinto redis backends to an external repository.


kinto_redis/tests/test_cache.py
import redis
import unittest
from pyramid import testing
from unittest import mock

from kinto.core.cache.testing import CacheTest
from kinto_redis import cache as redis_backend


class RedisCacheTest(CacheTest, unittest.TestCase):
    backend = redis_backend
    settings = {"cache_url": "", "cache_pool_size": 10, "cache_prefix": ""}

    def setUp(self):
        super(RedisCacheTest, self).setUp()
        self.client_error_patcher = mock.patch.object(
            self.cache._client, "execute_command", side_effect=redis.RedisError
        )

    def test_config_is_taken_in_account(self):
        config = testing.setUp(settings=self.settings)
        config.add_settings({"cache_url": "redis://:secret@peer.loc:4444/7"})
        backend = self.backend.load_from_config(config)
        self.assertDictEqual(
            backend.settings,
            {"host": "peer.loc", "password": "secret", "db": 7, "port": 4444},
        )

    def test_timeout_is_passed_to_redis_client(self):
        config = testing.setUp(settings=self.settings)
        config.add_settings({"cache_pool_timeout": "1.5"})
        backend = self.backend.load_from_config(config)
        self.assertEqual(backend._client.connection_pool.timeout, 1.5)


kinto_redis/tests/test_storage.py
from unittest import mock
import redis
import unittest
from pyramid import testing

from kinto.core.storage import exceptions
from kinto.core.storage.testing import StorageTest

from kinto_redis import storage as redisbackend


class RedisStorageTest(StorageTest, unittest.TestCase):
    backend = redisbackend
    settings = {"storage_pool_size": 50, "storage_url": ""}

    def setUp(self):
        super(RedisStorageTest, self).setUp()
        self.client_error_patcher = mock.patch.object(
            self.storage._client.connection_pool,
            "get_connection",
            side_effect=redis.exceptions.RedisError("connection error"),
        )

    def test_config_is_taken_in_account(self):
        config = testing.setUp(settings=self.settings)
        config.add_settings({"storage_url": "redis://:blah@store.loc:7777/6"})
        backend = self.backend.load_from_config(config)
        self.assertDictEqual(
            backend.settings,
            {"host": "store.loc", "password": "blah", "db": 6, "port": 7777},
        )

    def test_timeout_is_passed_to_redis_client(self):
        config = testing.setUp(settings=self.settings)
        config.add_settings({"storage_pool_timeout": "1.5"})
        backend = self.backend.load_from_config(config)
        self.assertEqual(backend._client.connection_pool.timeout, 1.5)

    def test_backend_error_is_raised_anywhere(self):
        with mock.patch.object(
            self.storage._client, "pipeline", side_effect=redis.RedisError
        ):
            StorageTest.test_backend_error_is_raised_anywhere(self)

    def test_list_all_handle_expired_values(self):
        obj = '{"id": "foo"}'.encode("utf-8")
        mocked_smember = mock.patch.object(
            self.storage._client, "smembers", return_value=["a", "b"]
        )
        mocked_mget = mock.patch.object(
            self.storage._client, "mget", return_value=[obj, None]
        )
        with mocked_smember:
            with mocked_mget:
                self.storage.list_all(**self.storage_kw)  # not raising

    def test_errors_logs_stack_trace(self):
        self.client_error_patcher.start()

        with mock.patch("kinto.core.storage.logger.exception") as exc_handler:
            with self.assertRaises(exceptions.BackendError):
                self.storage.list_all(**self.storage_kw)

        self.assertTrue(exc_handler.called)


kinto_redis/tests/test_listeners.py
import json
import uuid
import unittest
from datetime import datetime
from contextlib import contextmanager
from unittest import mock

from pyramid import testing

from kinto.core import initialization
from kinto.core.events import ResourceChanged, ACTIONS

from kinto_redis.storage import create_from_config


@contextmanager
def broken_redis():
    from redis import StrictRedis

    old = StrictRedis.lpush

    def push(*args, **kwargs):
        raise Exception("boom")

    StrictRedis.lpush = push
    yield
    StrictRedis.lpush = old


UID = str(uuid.uuid4())


class Resource(object):
    record_id = UID
    timestamp = 123456789


class ViewSet(object):
    def get_name(*args, **kw):
        return "collection"


class Service(object):
    viewset = ViewSet()


class Match(object):
    cornice_services = {"watev": Service()}
    pattern = "watev"


class Request(object):
    path = "/1/bucket/collection/"
    prefixed_userid = "tarek"
    matchdict = {"id": UID}
    registry = matched_route = Match()
    current_resource_name = "bucket"


class ListenerCalledTest(unittest.TestCase):
    def setUp(self):
        self.config = testing.setUp()
        self.config.add_settings(
            {"events_pool_size": 1, "events_url": "redis://localhost:6379/0"}
        )
        self._redis = create_from_config(self.config, prefix="events_")
        self._size = 0

        self.sample_event = ResourceChanged(
            {"action": ACTIONS.CREATE.value}, [], Request()
        )

    def _save_redis(self):
        self._size = self._redis.llen("kinto.core.events")

    def has_redis_changed(self):
        return self._redis.llen("kinto.core.events") > self._size

    def notify(self, event):
        self._save_redis()
        self.config.registry.notify(event)

    @contextmanager
    def redis_listening(self):
        config = self.config
        listener = "kinto_redis.listeners"

        # setting up the redis listener
        with mock.patch.dict(
            config.registry.settings,
            [("event_listeners", listener), ("event_listeners.redis.pool_size", "1")],
        ):
            initialization.setup_listeners(config)
            config.commit()
            yield

    def test_redis_is_notified(self):
        with self.redis_listening():
            # let's trigger an event
            self.notify(self.sample_event)
            self.assertTrue(self.has_redis_changed())

        # okay, we should have the first event in Redis
        last = self._redis.lpop("kinto.core.events")
        last = json.loads(last.decode("utf8"))
        self.assertEqual(last["action"], ACTIONS.CREATE.value)

    def test_notification_is_broken(self):
        with self.redis_listening():
            # an event with a bad JSON should silently break and send nothing
            # date time objects cannot be dumped
            event2 = ResourceChanged(
                {"action": ACTIONS.CREATE.value, "somedate": datetime.now()},
                [],
                Request(),
            )
            self.notify(event2)
            self.assertFalse(self.has_redis_changed())

    def test_redis_is_broken(self):
        with self.redis_listening():
            # if the redis call fails, same deal: we should ignore it
            self._save_redis()

            with broken_redis():
                self.config.registry.notify(self.sample_event)

            self.assertFalse(self.has_redis_changed())


kinto_redis/cache.py
from kinto.core.cache import CacheBase
from kinto_redis.storage import wrap_redis_error, create_from_config
from kinto.core.utils import json


class Cache(CacheBase):
    """Cache backend implementation using Redis.

    Enable in configuration::

        kinto.cache_backend = kinto_redis.cache

    *(Optional)* Instance location URI can be customized::

        kinto.cache_url = redis://localhost:6379/1

    A threaded connection pool is enabled by default::

        kinto.cache_pool_size = 50

    If the database is used for multiple Kinto deployement cache, you
    may want to add a prefix to every key to avoid collision::

        kinto.cache_prefix = stack1_

    :noindex:

    """

    def __init__(self, client, *args, **kwargs):
        super(Cache, self).__init__(*args, **kwargs)
        self._client = client

    @property
    def settings(self):
        return dict(self._client.connection_pool.connection_kwargs)

    def initialize_schema(self, dry_run=False):
        # Nothing to do.
        pass

    @wrap_redis_error
    def flush(self):
        self._client.flushdb()

    @wrap_redis_error
    def ttl(self, key):
        return self._client.ttl(self.prefix + key)

    @wrap_redis_error
    def expire(self, key, ttl):
        self._client.pexpire(self.prefix + key, int(ttl * 1000))

    @wrap_redis_error
    def set(self, key, value, ttl):
        if isinstance(value, bytes):
            raise TypeError("a string-like object is required, not 'bytes'")
        value = json.dumps(value)
        self._client.psetex(self.prefix + key, int(ttl * 1000), value)

    @wrap_redis_error
    def get(self, key):
        value = self._client.get(self.prefix + key)
        if value:
            value = value.decode("utf-8")
            return json.loads(value)

    @wrap_redis_error
    def delete(self, key):
        value = self.get(key)
        self._client.delete(self.prefix + key)
        return value


def load_from_config(config):
    settings = config.get_settings()
    client = create_from_config(config, prefix="cache_")
    return Cache(client, cache_prefix=settings["cache_prefix"])


kinto_redis/listeners.py
from kinto.core import logger
from kinto.core.listeners import ListenerBase
from kinto.core.utils import json

from kinto_redis.storage import create_from_config


class Listener(ListenerBase):
    """
    A Redis-based event listener that simply pushes the events payloads into
    the specified Redis list as they happen.

    This listener allows actions to be performed asynchronously, using Redis
    Pub/Sub notifications, or scheduled inspections of the queue.
    """

    def __init__(self, client, listname, *args, **kwargs):
        super(Listener, self).__init__(*args, **kwargs)
        self._client = client
        self.listname = listname

    def __call__(self, event):
        try:
            payload = json.dumps(event.payload)
        except TypeError:
            logger.error("Unable to dump the payload", exc_info=True)
            return
        try:
            self._client.lpush(self.listname, payload)
        except Exception:
            logger.error("Unable to send the payload to Redis", exc_info=True)


def load_from_config(config, prefix):
    settings = config.get_settings()
    settings.setdefault(prefix + "url", "")
    settings.setdefault(prefix + "pool_size", 25)
    listname = settings.get(prefix + "listname", "kinto.core.events")
    client = create_from_config(config, prefix)
    return Listener(client, listname=listname)


kinto_redis/storage.py
from functools import wraps

import redis
from urllib.parse import urlparse

from kinto.core.decorators import deprecate_kwargs
from kinto.core.storage import (
    exceptions,
    logger,
    DEFAULT_ID_FIELD,
    DEFAULT_MODIFIED_FIELD,
    DEFAULT_DELETED_FIELD,
)
from kinto.core.storage.memory import MemoryBasedStorage


def wrap_redis_error(func):
    @wraps(func)
    def wrapped(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except redis.exceptions.RedisError as e:
            logger.exception(e)
            raise exceptions.BackendError(original=e)

    return wrapped


def create_from_config(config, prefix=""):
    """Redis client instantiation from settings.
    """
    settings = config.get_settings()
    uri = settings[prefix + "url"]
    uri = urlparse(uri)
    kwargs = {
        "host": uri.hostname or "localhost",
        "port": uri.port or 6379,
        "password": uri.password or None,
        "db": int(uri.path[1:]) if uri.path else 0,
    }

    pool_size = settings.get(prefix + "pool_size")
    if pool_size is not None:
        kwargs["max_connections"] = int(pool_size)

    block_timeout = settings.get(prefix + "pool_timeout")
    if block_timeout is not None:
        kwargs["timeout"] = float(block_timeout)

    connection_pool = redis.BlockingConnectionPool(**kwargs)
    return redis.StrictRedis(connection_pool=connection_pool)


class Storage(MemoryBasedStorage):
    """Storage backend implementation using Redis.

    .. warning::

        Useful for very low server load, but won't scale since objects sorting
        and filtering are performed in memory.

    Enable in configuration::

        kinto.storage_backend = kinto_redis.storage

    *(Optional)* Instance location URI can be customized::

        kinto.storage_url = redis://localhost:6379/0

    A threaded connection pool is enabled by default::

        kinto.storage_pool_size = 50
    """

    def __init__(self, client, *args, readonly=False, **kwargs):
        super(Storage, self).__init__(*args, **kwargs)
        self.readonly = readonly
        self._client = client

    @property
    def settings(self):
        return dict(self._client.connection_pool.connection_kwargs)

    def _encode(self, obj):
        return self.json.dumps(obj)

    def _decode(self, obj):
        return self.json.loads(obj.decode("utf-8"))

    @wrap_redis_error
    def flush(self):
        self._client.flushdb()

    @wrap_redis_error
    def resource_timestamp(self, resource_name, parent_id):
        timestamp = self._client.get(
            "{0}.{1}.timestamp".format(resource_name, parent_id)
        )
        if timestamp:
            return int(timestamp)
        if self.readonly:
            error_msg = (
                "Cannot initialize empty resource timestamp when running in readonly."
            )
            raise exceptions.BackendError(message=error_msg)
        return self.bump_and_store_timestamp(resource_name, parent_id)

    @wrap_redis_error
    def bump_and_store_timestamp(
        self,
        resource_name,
        parent_id,
        obj=None,
        modified_field=None,
        last_modified=None,
    ):

        key = "{0}.{1}.timestamp".format(resource_name, parent_id)
        while 1:
            with self._client.pipeline() as pipe:
                try:
                    pipe.watch(key)
                    current_resource_timestamp = int(pipe.get(key) or 0)

                    current, resource_timestamp = self.bump_timestamp(
                        current_resource_timestamp, obj, modified_field, last_modified
                    )

                    pipe.multi()
                    pipe.set(key, resource_timestamp)
                    pipe.execute()
                    return current
                except redis.WatchError:  # pragma: no cover
                    # Our timestamp has been modified by someone else, let's
                    # retry.
                    # XXX: untested.
                    continue

    @deprecate_kwargs({"collection_id": "resource_name", "record": "obj"})
    @wrap_redis_error
    def create(
        self,
        resource_name,
        parent_id,
        obj,
        id_generator=None,
        id_field=DEFAULT_ID_FIELD,
        modified_field=DEFAULT_MODIFIED_FIELD,
        ignore_conflict=False,
    ):
        id_generator = id_generator or self.id_generator
        obj = self.json.loads(self.json.dumps(obj))
        if id_field in obj:
            # Raise unicity error if obj with same id already exists.
            try:
                existing = self.get(resource_name, parent_id, obj[id_field])
                raise exceptions.UnicityError(id_field, existing)
            except exceptions.ObjectNotFoundError:
                pass

        _id = obj.setdefault(id_field, id_generator())
        self.set_object_timestamp(
            resource_name, parent_id, obj, modified_field=modified_field
        )

        obj_key = "{0}.{1}.{2}.records".format(resource_name, parent_id, _id)
        with self._client.pipeline() as multi:
            multi.set(obj_key, self._encode(obj))
            multi.sadd("{0}.{1}.records".format(resource_name, parent_id), _id)
            multi.srem("{0}.{1}.deleted".format(resource_name, parent_id), _id)
            multi.execute()

        return obj

    @deprecate_kwargs({"collection_id": "resource_name"})
    @wrap_redis_error
    def get(
        self,
        resource_name,
        parent_id,
        object_id,
        id_field=DEFAULT_ID_FIELD,
        modified_field=DEFAULT_MODIFIED_FIELD,
    ):
        obj_key = "{0}.{1}.{2}.records".format(resource_name, parent_id, object_id)
        encoded_item = self._client.get(obj_key)
        if encoded_item is None:
            raise exceptions.ObjectNotFoundError(object_id)

        return self._decode(encoded_item)

    @deprecate_kwargs({"collection_id": "resource_name", "record": "obj"})
    @wrap_redis_error
    def update(
        self,
        resource_name,
        parent_id,
        object_id,
        obj,
        id_field=DEFAULT_ID_FIELD,
        modified_field=DEFAULT_MODIFIED_FIELD,
    ):
        obj = self.json.loads(self.json.dumps(obj))
        obj[id_field] = object_id
        self.set_object_timestamp(
            resource_name, parent_id, obj, modified_field=modified_field
        )

        obj_key = "{0}.{1}.{2}.records".format(resource_name, parent_id, object_id)
        with self._client.pipeline() as multi:
            multi.set(obj_key, self._encode(obj))
            multi.sadd("{0}.{1}.records".format(resource_name, parent_id), object_id)
            multi.srem("{0}.{1}.deleted".format(resource_name, parent_id), object_id)
            multi.execute()

        return obj

    @deprecate_kwargs({"collection_id": "resource_name"})
    @wrap_redis_error
    def delete(
        self,
        resource_name,
        parent_id,
        object_id,
        id_field=DEFAULT_ID_FIELD,
        with_deleted=True,
        modified_field=DEFAULT_MODIFIED_FIELD,
        deleted_field=DEFAULT_DELETED_FIELD,
        last_modified=None,
    ):
        obj_key = "{0}.{1}.{2}.records".format(resource_name, parent_id, object_id)
        with self._client.pipeline() as multi:
            multi.get(obj_key)
            multi.delete(obj_key)
            multi.srem("{0}.{1}.records".format(resource_name, parent_id), object_id)
            responses = multi.execute()

        encoded_item = responses[0]
        if encoded_item is None:
            raise exceptions.ObjectNotFoundError(object_id)

        existing = self._decode(encoded_item)

        # Need to delete the last_modified field.
        del existing[modified_field]

        self.set_object_timestamp(
            resource_name,
            parent_id,
            existing,
            modified_field=modified_field,
            last_modified=last_modified,
        )
        existing = self.strip_deleted_object(resource_name, parent_id, existing)

        if with_deleted:
            deleted_obj_key = "{0}.{1}.{2}.deleted".format(
                resource_name, parent_id, object_id
            )
            with self._client.pipeline() as multi:
                multi.set(deleted_obj_key, self._encode(existing))
                multi.sadd(
                    "{0}.{1}.deleted".format(resource_name, parent_id), object_id
                )
                multi.execute()

        return existing

    @deprecate_kwargs({"collection_id": "resource_name"})
    @wrap_redis_error
    def purge_deleted(
        self,
        resource_name,
        parent_id,
        before=None,
        id_field=DEFAULT_ID_FIELD,
        modified_field=DEFAULT_MODIFIED_FIELD,
    ):

        if resource_name is None:
            resource_name = "*"

        keys_pattern = "{0}.{1}.deleted".format(resource_name, parent_id)

        resources_keys = [
            key.decode("utf-8") for key in self._client.scan_iter(match=keys_pattern)
        ]

        resources_keys = [key for key in resources_keys if len(key.split(".")) == 3]
        with self._client.pipeline() as multi:
            for key in resources_keys:
                multi.smembers(key)
            results = multi.execute()

        number_deleted = 0
        for i, ids in enumerate(results):
            if len(ids) == 0:  # pragma: no cover
                continue

            resource_key = resources_keys[i]
            resource_name, parent_id, _ = resource_key.split(".")
            keys = [
                "{0}.{1}.{2}.deleted".format(
                    resource_name, parent_id, _id.decode("utf-8")
                )
                for _id in ids
            ]

            if len(keys) == 0:  # pragma: no cover
                continue

            encoded_results = self._client.mget(keys)
            deleted = [self._decode(r) for r in encoded_results if r]
            if before is not None:
                to_remove = [d["id"] for d in deleted if d[modified_field] < before]
            else:
                to_remove = [d["id"] for d in deleted]

            if len(to_remove) > 0:
                with self._client.pipeline() as pipe:
                    pipe.delete(
                        *[
                            "{0}.{1}.{2}.deleted".format(resource_name, parent_id, _id)
                            for _id in to_remove
                        ]
                    )
                    pipe.srem(resource_key, *to_remove)
                    pipe.execute()
            number_deleted += len(to_remove)
        return number_deleted

    def _get_objects_by_parent_id(self, parent_id, resource_name, with_meta=False):
        if resource_name is None:
            resource_name = "*"

        keys_pattern = "{0}.{1}.records".format(resource_name, parent_id)

        resources_keys = [
            key.decode("utf-8") for key in self._client.scan_iter(match=keys_pattern)
        ]

        resources_keys = [key for key in resources_keys if len(key.split(".")) == 3]
        with self._client.pipeline() as multi:
            for key in resources_keys:
                multi.smembers(key)
            results = multi.execute()

        objects = []
        for i, ids in enumerate(results):
            resource_name, parent_id, _ = resources_keys[i].split(".")

            if len(ids) == 0:  # pragma: no cover
                continue

            objects_keys = [
                "{0}.{1}.{2}.records".format(
                    resource_name, parent_id, _id.decode("utf-8")
                )
                for _id in ids
            ]
            results = self._client.mget(objects_keys)
            if with_meta:
                resource_objects = [
                    dict(
                        __resource_name__=resource_name,
                        __parent_id__=parent_id,
                        **self._decode(r)
                    )
                    for r in results
                    if r
                ]
            else:
                resource_objects = [self._decode(r) for r in results if r]
            objects.extend(resource_objects)

        return objects

    @wrap_redis_error
    def list_all(
        self,
        resource_name,
        parent_id,
        filters=None,
        sorting=None,
        pagination_rules=None,
        limit=None,
        include_deleted=False,
        id_field=DEFAULT_ID_FIELD,
        modified_field=DEFAULT_MODIFIED_FIELD,
        deleted_field=DEFAULT_DELETED_FIELD,
    ):

        objects = self._get_objects_by_parent_id(parent_id, resource_name)

        deleted = []
        if include_deleted:
            keys_pattern = "{0}.{1}.deleted".format(resource_name, parent_id)

            resources_keys = [
                key.decode("utf-8")
                for key in self._client.scan_iter(match=keys_pattern)
            ]

            resources_keys = [key for key in resources_keys if len(key.split(".")) == 3]
            with self._client.pipeline() as multi:
                for key in resources_keys:
                    multi.smembers(key)
                results = multi.execute()

            deleted = []
            for i, ids in enumerate(results):
                resource_name, parent_id, _ = resources_keys[i].split(".")

                if len(ids) == 0:  # pragma: no cover
                    continue

                deleted_keys = [
                    "{0}.{1}.{2}.deleted".format(
                        resource_name, parent_id, _id.decode("utf-8")
                    )
                    for _id in ids
                ]
                results = self._client.mget(deleted_keys)
                resource_objects = [self._decode(r) for r in results if r]
                deleted.extend(resource_objects)

        objects, _ = self.extract_object_set(
            objects + deleted,
            filters,
            sorting,
            id_field,
            deleted_field,
            pagination_rules,
            limit,
        )

        return objects

    @wrap_redis_error
    def count_all(
        self,
        resource_name,
        parent_id,
        filters=None,
        sorting=None,
        pagination_rules=None,
        limit=None,
        include_deleted=False,
        id_field=DEFAULT_ID_FIELD,
        modified_field=DEFAULT_MODIFIED_FIELD,
        deleted_field=DEFAULT_DELETED_FIELD,
    ):

        objects = self._get_objects_by_parent_id(parent_id, resource_name)

        _, count = self.extract_object_set(
            objects, filters, sorting, id_field, deleted_field, pagination_rules, limit
        )

        return count

    @deprecate_kwargs({"collection_id": "resource_name"})
    @wrap_redis_error
    def delete_all(
        self,
        resource_name,
        parent_id,
        filters=None,
        sorting=None,
        pagination_rules=None,
        limit=None,
        id_field=DEFAULT_ID_FIELD,
        with_deleted=True,
        modified_field=DEFAULT_MODIFIED_FIELD,
        deleted_field=DEFAULT_DELETED_FIELD,
    ):
        objects = self._get_objects_by_parent_id(
            parent_id, resource_name, with_meta=True
        )

        objects, count = self.extract_object_set(
            objects=objects,
            filters=filters,
            sorting=sorting,
            pagination_rules=pagination_rules,
            limit=limit,
            id_field=id_field,
            deleted_field=deleted_field,
        )

        deleted = [
            self.delete(
                r.pop("__resource_name__"),
                r.pop("__parent_id__"),
                r[id_field],
                id_field=id_field,
                with_deleted=with_deleted,
                modified_field=modified_field,
                deleted_field=deleted_field,
            )
            for r in objects
        ]
        return deleted


def load_from_config(config):
    client = create_from_config(config, prefix="storage_")
    return Storage(client)
